bin/manage.py

```diff
@@ -0,0 +1,470 @@
+
+# ---------------------------------------------------------
+
+#class ContainerPilot(object):
+#    """
+#    ContainerPilot config is where we rewrite ContainerPilot's own config
+#    so that we can dynamically alter what service we advertise
+#    """
+#
+#    def __init__(self):
+#        # TODO: we should make sure we can support JSON-in-env-var
+#        # the same as ContainerPilot itself
+#        self.path = get_environ('CONTAINERPILOT', None).replace('file://', '')
+#        with open(self.path, 'r') as f:
+#            self.config = json.loads(f.read())
+#
+#    @debug
+#    def update(self, state):
+#        if state and self.config['services'][0]['name'] != state:
+#            self.config['services'][0]['name'] = state
+#            self.render()
+#            return True
+#
+#    @debug
+#    def render(self):
+#        new_config = json.dumps(self.config)
+#        with open(self.path, 'w') as f:
+#            f.write(new_config)
+#
+#    def reload(self):
+#        """ force ContainerPilot to reload its configuration """
+#        log.info('Reloading ContainerPilot configuration.')
+#        os.kill(1, signal.SIGHUP)
+
+# ---------------------------------------------------------
+# Top-level functions called by ContainerPilot or forked by this program
+
+@debug
+def pre_start():
+    """
+    MongoDB must be running in order to execute most of our setup behavior
+    """
+    # TODO is there anything that needs to be done before starting mongo?
+    sys.exit(0)
+
+@debug
+def pre_stop():
+    """
+    If we are the primary in the MongoDB replica set, we need to step down
+    because we are about to be shut down
+    """
+
+    ip = get_ip()
+    local_mongo = MongoClient(ip, connect=False)
+
+    # since we are shutting down, it is ok to stop if mongo is already non-responsive
+    if not is_mongo_up(local_mongo):
+        return True
+
+    try:
+        repl_status = local_mongo.admin.command('replSetGetStatus')
+        is_mongo_primary = repl_status['myState'] == 1
+        # ref https://docs.mongodb.com/manual/reference/replica-states/
+    except Exception as e:
+        log.error(e, 'unable to get primary status while shuting down')
+        return True
+
+    if is_mongo_primary:
+        # the primary will wait up to X seconds for a secondary member
+        # to catch up and sets itself as ineligible to be primary again for 60 seconds
+        # https://docs.mongodb.com/manual/reference/command/replSetStepDown/i
+        # this is set to 8 so that we timeout before `docker stop` would send a sigkill
+        try:
+            # TODO stepdown times in ENV vars
+            local_mongo.admin.command('replSetStepDown', 60, secondaryCatchUpPeriodSecs=8)
+        except ExecutionTimeout as e:
+            # stepdown fails, ie no secondary that is caught up
+            log.debug(e)
+            try:
+                # force
+                local_mongo.admin.command('replSetStepDown', 60, force=True)
```

Comment:
> Or maybe recommend higher timeouts to docker stop, and containerpliot?

Keep in mind that lots of people will be using some kind of higher-level scheduler so telling them not to use `docker stop` might not be relevant. Setting the timeouts is the right way to go.

> Can we to adjust the stopTimeout in containerpilot from the python so that it can be set via env?

You can interpolate it in the ContainerPilot config from the env. See "template configuration" at the bottom of https://www.joyent.com/containerpilot/docs/configuration